pubsub/natspubsub/nats.go

```diff
@@ -0,0 +1,263 @@
+	return &subscription{nc, sub, err}
+}
+
+// ReceiveBatch implements driver.ReceiveBatch.
+func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	if s == nil || s.nsub == nil {
+		return nil, nats.ErrBadSubscription
+	}
+
+	// Return right away if the ctx has an error.
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+	// Just in case
+	if maxMessages == 0 {
+		maxMessages = 1
+	}
+
+	var ms []*driver.Message
+
+	// We will assume the desired goal is at least one message since the public API only has Receive().
+	// We will load all the messages that are already queued up first.
+	for {
+		msg, err := s.nsub.NextMsg(0)
+		if err == nil {
+			dm, err := decode(msg)
+			if err != nil {
+				return nil, err
+			}
+			ms = append(ms, dm)
+			if len(ms) >= maxMessages {
+				break
+			}
+		} else if err == nats.ErrTimeout {
+			break
+		} else {
+			return nil, err
+		}
+	}
+
+	// If we have anything go ahead and return here.
+	if len(ms) > 0 {
+		return ms, nil
+	}
+
+	// Here we will assume the ctx has a deadline and will allow it to do its thing. We may
+	// get a burst of messages but we will only wait once here and let the next call grab them.
+	// The reason is so that if deadline is not properly set we will not needlessly wait here
+	// for more messages when the user most likely only wants one.
+	msg, err := s.nsub.NextMsgWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	dm, err := decode(msg)
+	if err != nil {
+		return nil, err
+	}
+	ms = append(ms, dm)
+	return ms, nil
+}
+
+// Convert NATS msgs to *driver.Message.
+func decode(msg *nats.Msg) (*driver.Message, error) {
+	if msg == nil {
+		return nil, nats.ErrInvalidMsg
+	}
+	var dm driver.Message
+	// Everything is in the msg.Data
+	dec := codec.NewDecoderBytes(msg.Data, &mh)
+	defer dec.Release()
+
+	dec.Decode(&dm)
+	dm.AckID = -1 // Not applicable to NATS
+	dm.AsFunc = messageAsFunc(msg)
+
+	return &dm, nil
+}
+
+func messageAsFunc(msg *nats.Msg) func(interface{}) bool {
+	return func(i interface{}) bool {
+		p, ok := i.(**nats.Msg)
+		if !ok {
+			return false
+		}
+		*p = msg
+		return true
+	}
+}
+
+// SendAcks implements driver.Subscription.SendAcks. NATS does not need Acks since
+// it is At-Most-Once QoS.
+func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
+	return nil
+}
+
+// IsRetryable implements driver.Subscription.IsRetryable.
+func (s *subscription) IsRetryable(error) bool { return false }
+
+// As implements driver.Subscription.As.
+func (s *subscription) As(i interface{}) bool {
+	c, ok := i.(**nats.Subscription)
+	if !ok {
+		return false
+	}
+	*c = s.nsub
+	return true
+}
+
+// ErrorAs implements driver.Subscription.ErrorAs
+func (*subscription) ErrorAs(error, interface{}) bool {
+	return false
+}
+
+// ErrorCode implements driver.Subscription.ErrorCode
+func (*subscription) ErrorCode(error) gcerrors.ErrorCode { return gcerrors.Unknown }
```

Comment:
This should most likely return gcerrrors.FailedPrecondition for the error of the subscription or its connection being nil.

Reply:
Will take a look.